### DISCUSSION TITLE
Cannot gather fact USER using docker

### DISCUSSION THREAD

**Discussion by staticdev** [ansible-community/molecule-plugins#26] [docker] [answered]
I have a working playbook that when run in molecule-docker fails to get ansible_env.USER. When I run in verbose mode I see [Gathering Facts] runs on converge successfully and get other variables, but not USER.

When I run the same playbook in my machine it gets the user. Does anyone know the reason for this difference? If so, how should I test it with molecule?

If interested in checking the playbook, the code with molecule is here: https://github.com/staticdev/linux-workstation-playbook/pull/316

**Comment by staticdev** [comment]
@ssbarnea maybe you have an idea on this one?

**Reply by catarse-natural** [reply]
I have the exact same problem, did you find a solution @staticdev ?

**Reply by staticdev** [reply]
I am sorry, but not yet =/

**Comment by zhan9san** [comment, marked as answer]
If you have to use `{{ ansible_env.USER }}`, both work for your case.

1. Try `"{{ ansible_facts.user_id }}"` in your role.
2. Add an environment variable(USER: root) in `molecule.yml` as below.

  ```yaml
  ---
  dependency:
    name: galaxy
  driver:
    name: docker
  platforms:
    - name: instance
      image: xxx
      # image: x
      pre_build_image: true
      env:
        USER: root
  ```

> When I run the same playbook in my machine it gets the user. Does anyone know the reason for this difference?

It's because the [ansible connection](https://docs.ansible.com/ansible/latest/plugins/connection.html) is different.

When you run playbook in real target server, the connection plugin must be `ssh` while when in `molecule` with `docker` plugin, the connection is `community.docker.docker`. By default, the user running in container is `root`, but this user is not set as an environment variable in container.

When using `community.docker.docker` connection, even `gather_facts` is enabled, as no `USER` environment variable in container, no `ansible_env.USER`(AKA `ansible_facts.env.USER`) would be set.

---

No `USER` environment variable as below.

```shell
$ docker run --rm --entrypoint env geerlingguy/docker-debian11-ansible:latest
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=dd4d52ad805a
pip_packages=ansible cryptography
HOME=/root
```

As we can see, no `USRER` in environment variables.

**Reply by staticdev** [reply]
@zhan9san thanks a lot for the answer and explanation. I tried the first option, not sure I understood correctly, just replaced `{{ ansible_env.USER }}` with `{{ ansible_facts.user_id }}` but I believe this is not what you meant. The second solution though worked great!

**Reply by zhan9san** [reply]
> not sure I understood correctly, just replaced {{ ansible_env.USER }} with {{ ansible_facts.user_id }} but I believe this is not what you meant.

It's what I mean.

technically, the value of  {{ ansible_facts.user_id }} is what you want in both your real playbook and converge playbook.

In you real playbook, it's ssh connection user.
In docker connection, it would be set to root by default.

Didn't work?

**Reply by staticdev** [reply]
@zhan9san you are completely right, I am sorry! I double checked and the problem was I had it in multiple points of my playbook (so maybe setting the ENV is the easier solution). Thanks a lot again.
